File: packages/crash-detection/ios/SentianceCrashDetection/SentianceCrashDetection.m
//
//  SentianceCrashDetection.h
//  SentianceCrashDetection
//
//  Created by Mohammed Aouf Zouag on 18/03/2025.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(SentianceCrashDetection, NSObject)

RCT_EXTERN_METHOD(isVehicleCrashDetectionSupported:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(invokeDummyVehicleCrash:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

@end

File: packages/crash-detection/ios/SentianceCrashDetection/SentianceCrashDetection.swift
//
//  SentianceCrashDetection.swift
//  SentianceCrashDetection
//
//  Created by Mohammed Aouf Zouag on 18/03/2025.
//

import Foundation
import SENTSDK

@objc(SentianceCrashDetection)
class SentianceCrashDetection: NSObject {

    private let sdkNotInitializedError = (code: ESDKNotInitialized, message: "SDK is not initialized")
    private let sentiance = Sentiance.shared

    @objc
    func isVehicleCrashDetectionSupported(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard ensureSDKInitialized(reject) else { return }
        resolve(sentiance.isVehicleCrashDetectionSupported)
    }

    @objc
    func invokeDummyVehicleCrash(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        guard ensureSDKInitialized(reject) else { return }
        sentiance.invokeDummyVehicleCrash()
        resolve(true)
    }

    private func ensureSDKInitialized(_ reject: RCTPromiseRejectBlock) -> Bool {
        guard sentiance.initState == SENTSDKInitState.initialized else {
            reject(sdkNotInitializedError.code, sdkNotInitializedError.message, nil)
            return false
        }
        return true
    }

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }
}
